frontend/public/module/k8s/swagger.ts

```diff
@@ -0,0 +1,80 @@
+import * as _ from 'lodash-es';
+
+import { STORAGE_PREFIX } from '../../const';
+import { coFetchJSON } from '../../co-fetch';
+import { SwaggerDefinitions } from './';
+
+const SWAGGER_LOCAL_STORAGE_KEY = `${STORAGE_PREFIX}/swagger-definitions`;
+const SWAGGER_TIMESTAMP_LOCAL_STORAGE_KEY = `${STORAGE_PREFIX}/swagger-last-updated`;
+
+export const getStoredSwagger = (): SwaggerDefinitions => {
+  const json = window.localStorage.getItem(SWAGGER_LOCAL_STORAGE_KEY);
+  if (!json) {
+    return null;
+  }
+  try {
+    return JSON.parse(json);
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.error('Could not parse swagger JSON.', e);
+    return null;
+  }
+};
+
+const isSwaggerStale = () => {
+  // Avoid refreshing the entire document on cellular connections.
+  const connection = (navigator as any).connection || (navigator as any).mozConnection || (navigator as any).webkitConnection;
+  if (connection && connection.effectiveType === 'cellular') {
+    return false;
+  }
+  // Store the swagger definitions in localStorage, but refresh once per day.
+  // This document is large, so avoid reloading it too often.
+  const lastUpdated = Number(window.localStorage.getItem(SWAGGER_TIMESTAMP_LOCAL_STORAGE_KEY));
+  return !_.isFinite(lastUpdated) || lastUpdated < (Date.now() - (1000 * 60 * 60 * 24));
+};
+
+const storeSwagger = (swagger: SwaggerAPISpec) => {
+  // Only store definitions to reduce the document size.
+  const json = JSON.stringify(swagger.definitions || {});
+  window.localStorage.setItem(SWAGGER_LOCAL_STORAGE_KEY, json);
+  window.localStorage.setItem(SWAGGER_TIMESTAMP_LOCAL_STORAGE_KEY, `${Date.now()}`);
+};
+
+export const fetchSwagger = async(): Promise<SwaggerDefinitions> => {
+  try {
+    const storedSwagger = getStoredSwagger();
+    if (storedSwagger && !isSwaggerStale()) {
+      return getStoredSwagger();
```

Comment:
Looks like we are parsing the JSON twice here. Should be able to just return the JSON that was parsed on L45:

```suggestion
      return storedSwagger;
```